test/tree/general/base.py

```diff
@@ -144,13 +144,40 @@ def test_numeric_matmul(self):
                 }
             })
 
-            original_id = id(t1._detach())
-            original_id_x = id(t1.x._detach())
-            t4 = t1 @ t2
-            t1 @= t2
-            assert npeq(t1, t4)
-            assert id(t1._detach()) == original_id
-            assert id(t1.x._detach()) == original_id_x
+            class Ft:
+                def __init__(self, x):
+                    self.x = x
+                    self.c = 0
+                    self.rc = 0
+                    self.ic = 0
+
+                def __matmul__(self, other):
+                    self.c += 1
+                    return Ft(self.x @ other.x)
+
+                def __rmatmul__(self, other):
+                    self.rc += 1
+                    return Ft(other.x @ self.x)
+
+                def __imatmul__(self, other):
+                    self.ic += 1
+                    self.x = self.x @ other.x
+                    return self
+
+            tt1 = mapping(t1, lambda x, p: Ft(x))
+            tt2 = mapping(t2, lambda x, p: Ft(x))
+
+            original_id = id(tt1._detach())
+            original_id_x = id(tt1.x._detach())
+            tt4 = tt1 @ tt2
+            tt1 @= tt2
+            assert npeq(tt1.x, tt4.x)
```

Comment:
rename `npeq`, which may be confused

Reply:
Renamed from `npeq` to `tnp_array_equal`.